docs/proposals/message-proto-design.md

```diff
@@ -0,0 +1,346 @@
+ ```
+ RegisterType("", "string.content")
+ ``` 
+if we want bool types can conversion to Any types. we need to add
+```
+RegisterType(true, "bool.content")
+```
+
+if we want v1.Pod types ("k8s.io/api/core/v1") can conversion to Any types. we need to add
+``` 
+RegisterType(v1.Pod{}, "v1.pod.content")
+```
+the second parameter used to generate TypeUrl of `Any` type. the TypeUrl must be unique, otherwise it will panic.
+
+* `func MarshalAny(v interface{}) (*types.Any, error) `,  this function takes interface and encodes it into google.protobuf.Any. and this interface must be Registerd by `RegisterType` function.
+* `func UnmarshalAny(any *types.Any) (interface{}, error)`, this function parses the google.protobuf.Any message into a interface.
+
+## Consider compatibility  
+
+In order to be more compatible, i try to avoid modifying methods of message. Only the following method has been modified, but I don't think it will have much impact。
+```
+from:
+func (msg *Message) BuildRouter(source, group, res, opr string) *Message {
+to:
+func (msg *Message) BuildRouter(source, group string, res interface{}, opr string) *Message {
+```
+
+```
+from
+func (msg *Message) SetResourceOperation(res, opr string) *Message {
+to
+func (msg *Message) SetResourceOperation(res interface{}, opr string) *Message {
+```
+
+```
+from:
+func (msg *Message) GetResource() string {
+to
+func (msg *Message) GetResource() interface{} {
+```
+
+let me just mention `FillBody()` function again:
+```
+func (msg *Message) FillBody(content interface{}) *Message {
+	msg.Content = content
+	real, err := typeurl.MarshalAny(content)
+	if err != nil {
+		panic(errors.Errorf("Marshal content interface to any type error %v", err))
+	}
+	msg.Data = real
+	return msg
+}
+```
+The parameter `content interface{}` is not changed, but the parameter passed in must be registered by `RegisterType()`.
+so if you want to fill a map or a struct as message\`s Data, the map or struct must registerd by  `RegisterType()`. 
+in `pkg/core/typeurl` package of beehive project, it has registerd some frequently used types.
+```
+func init() {
+	registry = make(map[reflect.Type]string)
+	revRegistry = make(map[string]reflect.Type)
+	lock = sync.Mutex{}
+	// Register sting
+	RegisterType("", "string.content")
+	// Register boolean
+	RegisterType(true, "bool.content")
+	// register slice
+	RegisterType([]interface{}{}, "slice.interface.content")
+	// register pod
+	RegisterType(v1.Pod{}, "v1.pod.content")
+}
+```
+
+# How to use in kubeedge project
+
+take `DownstreamController` as an example.
+
+In `DownstreamController` , it will send 6 kind of message. `Pod`, `CongfigMap`,`secret`,`node`,`service`,`endpoint`. They have similar resource formats(`/node/{nodeID}/{namespace}/{resourceType}/{resource}`).
+
+so we need define a interface :
+```
+type ToEdgeResourcer interface {
```

Comment:
why do we need an interface ?

Reply:
yes, you are right, for this example,  interface is not necessary. we can only define a  particular structure:
```
type ToEdgeResource struct {
	ResourceType string
	NameSpace    string
	NodeID       string
	ID           string
}
```
and let it registerd by `RegisterType()` function.

Reply:
can we make these changes in the design ?

Reply:
ok